In [ ]:
!pip install apyori

  Preparing metadata (setup.py) ... done
  Created wheel for apyori: filename=apyori-1.1.2-py3-none-any.whl size=5955 sha256=2d51d9040bfc18dddfb6855d1a461c64154feb458ef486a7b1e9306c60ec2b58
  Stored in directory: /root/.cache/pip/wheels/c4/1a/79/20f55c470a50bb3702a8cb7c94d8ada15573538c7f4baebe2d
Successfully built apyori


In [ ]:
import pandas as pd
import numpy as np
from apyori import apriori

In [ ]:
df = pd.read_csv('/content/groceries_dataset.csv')

In [ ]:
records = []
for i in range(0, 9835):
    records.append([str(df.values[i,u]) for u in range(0, 32)])

In [ ]:
rules = apriori(records, min_support=0.0022, min_confidence=0.20, min_lift=3, min_length = 4)

In [ ]:
# Store rules in result variable
results = list(rules)# See the items that were bought together with their support
results_list = []
for i in range(0, len(results)):
    results_list.append('RULE:\t' + str(results[i][0]) + '\nSUPPORT:\t' + str(results[i][1]))

In [ ]:
print(len(results_list))

1072


In [ ]:
print(results_list)

["RULE:\tfrozenset({'Instant food products', 'hamburger meat'})\nSUPPORT:\t0.003050330452465684", "RULE:\tfrozenset({'baking powder', 'whipped/sour cream'})\nSUPPORT:\t0.004575495678698526", "RULE:\tfrozenset({'beef', 'root vegetables'})\nSUPPORT:\t0.017386883579054397", "RULE:\tfrozenset({'berries', 'whipped/sour cream'})\nSUPPORT:\t0.009049313675648195", "RULE:\tfrozenset({'liquor', 'bottled beer'})\nSUPPORT:\t0.004677173360447382", "RULE:\tfrozenset({'red/blush wine', 'bottled beer'})\nSUPPORT:\t0.004880528723945094", "RULE:\tfrozenset({'condensed milk', 'coffee'})\nSUPPORT:\t0.002541942043721403", "RULE:\tfrozenset({'margarine', 'flour'})\nSUPPORT:\t0.0037620742247076767", "RULE:\tfrozenset({'sugar', 'flour'})\nSUPPORT:\t0.00498220640569395", "RULE:\tfrozenset({'whipped/sour cream', 'flour'})\nSUPPORT:\t0.004067107269954245", "RULE:\tfrozenset({'frankfurter', 'mustard'})\nSUPPORT:\t0.002541942043721403", "RULE:\tfrozenset({'herbs', 'root vegetables'})\nSUPPORT:\t0.00701576004067107

RULE 1:\tfro
zenset({'Instant food products', 'hamburger meat'})\nSUPPORT:\t0.003050330452465684

If a customer buys "Instant food products" and "hamburger meat", then there is a support of 0.00305, which means that these items are purchased together in approximately 0.305% of all transactions.

In [ ]:
columns=["age","workclass","fnlwgt","education","education-num","marital-status","occupation","relationship","race","sex","capital-gain","capital-loss","hours-per-week","native-country","income"]
df = pd.read_csv('/content/DATA.csv',header=None,names=columns)
df.head()

,age,workclass,fnlwgt,education,education-num,marital-status,occupation,relationship,race,sex,capital-gain,capital-loss,hours-per-week,native-country,income
0,39,State-gov,77516,Bachelors,13,Never-married,Adm-clerical,Not-in-family,White,Male,2174,0,40,United-States,<=50K
1,50,Self-emp-not-inc,83311,Bachelors,13,Married-civ-spouse,Exec-managerial,Husband,White,Male,0,0,13,United-States,<=50K
2,38,Private,215646,HS-grad,9,Divorced,Handlers-cleaners,Not-in-family,White,Male,0,0,40,United-States,<=50K
3,53,Private,234721,11th,7,Married-civ-spouse,Handlers-cleaners,Husband,Black,Male,0,0,40,United-States,<=50K
4,28,Private,338409,Bachelors,13,Married-civ-spouse,Prof-specialty,Wife,Black,Female,0,0,40,Cuba,<=50K


In [ ]:
df.isnull().sum()

age               0
workclass         0
fnlwgt            0
education         0
education-num     0
marital-status    0
occupation        0
relationship      0
race              0
sex               0
capital-gain      0
capital-loss      0
hours-per-week    0
native-country    0
income            0
dtype: int64

In [ ]:
df.columns

Index(['age', 'workclass', 'fnlwgt', 'education', 'education-num',
       'marital-status', 'occupation', 'relationship', 'race', 'sex',
       'capital-gain', 'capital-loss', 'hours-per-week', 'native-country',
       'income'],
      dtype='object')

In [ ]:
num_cols = df.select_dtypes(include=['int64', 'float64', 'int32']).columns
num_cols

Index(['age', 'fnlwgt', 'education-num', 'capital-gain', 'capital-loss',
       'hours-per-week'],
      dtype='object')

In [ ]:
# One-Hot Encoding
from sklearn.preprocessing import OneHotEncoder

categorical_cols = ['workclass', 'education', 'marital-status', 'occupation', 'relationship', 'race', 'sex', 'native-country']

# Create a new DataFrame with only relevant columns
categorical_df = df[categorical_cols]
num_df=df[num_cols]
encoder = OneHotEncoder()

# Fit and transform the categorical data
encoded_data = encoder.fit_transform(categorical_df).toarray()

# Create a DataFrame from the encoded data
encoded_df = pd.DataFrame(encoded_data, columns=encoder.get_feature_names_out(categorical_cols))

# Combine the encoded DataFrame with the original numeric columns
binarized_df = pd.concat([encoded_df], axis=1)

# Display the combined DataFrame
print(binarized_df.head())

   workclass_ ?  workclass_ Federal-gov  workclass_ Local-gov  \
0           0.0                     0.0                   0.0   
1           0.0                     0.0                   0.0   
2           0.0                     0.0                   0.0   
3           0.0                     0.0                   0.0   
4           0.0                     0.0                   0.0   

   workclass_ Never-worked  workclass_ Private  workclass_ Self-emp-inc  \
0                      0.0                 0.0                      0.0   
1                      0.0                 0.0                      0.0   
2                      0.0                 1.0                      0.0   
3                      0.0                 1.0                      0.0   
4                      0.0                 1.0                      0.0   

   workclass_ Self-emp-not-inc  workclass_ State-gov  workclass_ Without-pay  \
0                          0.0                   1.0                     0.0  

In [ ]:
from mlxtend.frequent_patterns import apriori
from mlxtend.frequent_patterns import association_rules

# Set your support and confidence thresholds
min_support_values = [0.05, 0.1, 0.15]
min_confidence = 0.7

# Calculate frequent itemsets
for min_support in min_support_values:
    frequent_itemsets = apriori(binarized_df, min_support=min_support, use_colnames=True)

    # Extract association rules
    rules = association_rules(frequent_itemsets, metric="confidence", min_threshold=min_confidence)

    # Rank the top 10 rules
    top_10_rules = rules.nlargest(10, "confidence")

    # Print rules and explanations
    print(f"Rules for min_support={min_support:.2f}:")
    for i, (_, row) in enumerate(top_10_rules.iterrows(), start=1):
        antecedent = row['antecedents']
        consequent = row['consequents']
        confidence = row['confidence']

        print(f"Rule {i}: If {antecedent} then {consequent} (Confidence: {confidence:.2f})")

/usr/local/lib/python3.10/dist-packages/mlxtend/frequent_patterns/fpcommon.py:110: DeprecationWarning: DataFrames with non-bool types result in worse computationalperformance and their support might be discontinued in the future.Please use a DataFrame with bool type
  warnings.warn(


Rules for min_support=0.05:
Rule 1: If frozenset({'workclass_ ?'}) then frozenset({'occupation_ ?'}) (Confidence: 1.00)
Rule 2: If frozenset({'workclass_ ?', 'native-country_ United-States'}) then frozenset({'occupation_ ?'}) (Confidence: 1.00)
Rule 3: If frozenset({'education_ Bachelors', 'relationship_ Husband'}) then frozenset({'sex_ Male'}) (Confidence: 1.00)
Rule 4: If frozenset({'relationship_ Husband', 'education_ Some-college'}) then frozenset({'sex_ Male'}) (Confidence: 1.00)
Rule 5: If frozenset({'occupation_ Exec-managerial', 'relationship_ Husband'}) then frozenset({'marital-status_ Married-civ-spouse'}) (Confidence: 1.00)
Rule 6: If frozenset({'occupation_ Craft-repair', 'relationship_ Husband'}) then frozenset({'sex_ Male'}) (Confidence: 1.00)
Rule 7: If frozenset({'occupation_ Exec-managerial', 'relationship_ Husband'}) then frozenset({'sex_ Male'}) (Confidence: 1.00)
Rule 8: If frozenset({'occupation_ Prof-specialty', 'relationship_ Husband'}) then frozenset({'sex_ Male

/usr/local/lib/python3.10/dist-packages/mlxtend/frequent_patterns/fpcommon.py:110: DeprecationWarning: DataFrames with non-bool types result in worse computationalperformance and their support might be discontinued in the future.Please use a DataFrame with bool type
  warnings.warn(


Rules for min_support=0.10:
Rule 1: If frozenset({'relationship_ Husband'}) then frozenset({'sex_ Male'}) (Confidence: 1.00)
Rule 2: If frozenset({'marital-status_ Married-civ-spouse', 'relationship_ Husband'}) then frozenset({'sex_ Male'}) (Confidence: 1.00)
Rule 3: If frozenset({'relationship_ Husband', 'race_ White'}) then frozenset({'sex_ Male'}) (Confidence: 1.00)
Rule 4: If frozenset({'marital-status_ Married-civ-spouse', 'relationship_ Husband', 'race_ White'}) then frozenset({'sex_ Male'}) (Confidence: 1.00)
Rule 5: If frozenset({'relationship_ Husband', 'native-country_ United-States'}) then frozenset({'sex_ Male'}) (Confidence: 1.00)
Rule 6: If frozenset({'marital-status_ Married-civ-spouse', 'relationship_ Husband', 'native-country_ United-States'}) then frozenset({'sex_ Male'}) (Confidence: 1.00)
Rule 7: If frozenset({'relationship_ Husband', 'native-country_ United-States', 'race_ White'}) then frozenset({'sex_ Male'}) (Confidence: 1.00)
Rule 8: If frozenset({'marital-stat

/usr/local/lib/python3.10/dist-packages/mlxtend/frequent_patterns/fpcommon.py:110: DeprecationWarning: DataFrames with non-bool types result in worse computationalperformance and their support might be discontinued in the future.Please use a DataFrame with bool type
  warnings.warn(


In [ ]:
from mlxtend.frequent_patterns import apriori
from mlxtend.frequent_patterns import association_rules

def calculate_interest(rule):
    return rule['confidence'] * rule['lift']

min_support_values = [0.05, 0.1, 0.15]
min_interest_values = [0.5, 0.7, 0.9]

for min_support in min_support_values:
    frequent_itemsets = apriori(binarized_df, min_support=min_support, use_colnames=True)

    for min_interest in min_interest_values:
        rules = association_rules(frequent_itemsets, metric="confidence", min_threshold=min_confidence)

        # Calculate interest measure for each rule
        rules['interest'] = rules.apply(calculate_interest, axis=1)

        # Rank the top 10 rules based on interest
        top_10_rules = rules.nlargest(10, "interest")

        # Print rules and explanations
        print(f"Rules for min_support={min_support:.2f} and min_interest={min_interest:.2f}:")
        for i, (_, row) in enumerate(top_10_rules.iterrows(), start=1):
            antecedent = row['antecedents']
            consequent = row['consequents']
            confidence = row['confidence']
            lift = row['lift']
            interest = row['interest']

            print(f"Rule {i}: If {antecedent} then {consequent} (Confidence: {confidence:.2f}, Lift: {lift:.2f}, Interest: {interest:.2f})")

/usr/local/lib/python3.10/dist-packages/ipykernel/ipkernel.py:283: DeprecationWarning: `should_run_async` will not call `transform_cell` automatically in the future. Please pass the result to `transformed_cell` argument and any exception that happen during thetransform in `preprocessing_exc_tuple` in IPython 7.17 and above.
  and should_run_async(code)
/usr/local/lib/python3.10/dist-packages/mlxtend/frequent_patterns/fpcommon.py:110: DeprecationWarning: DataFrames with non-bool types result in worse computationalperformance and their support might be discontinued in the future.Please use a DataFrame with bool type
  warnings.warn(


Rules for min_support=0.05 and min_interest=0.50:
Rule 1: If frozenset({'workclass_ ?'}) then frozenset({'occupation_ ?'}) (Confidence: 1.00, Lift: 17.67, Interest: 17.67)
Rule 2: If frozenset({'workclass_ ?', 'native-country_ United-States'}) then frozenset({'occupation_ ?'}) (Confidence: 1.00, Lift: 17.67, Interest: 17.67)
Rule 3: If frozenset({'occupation_ ?'}) then frozenset({'workclass_ ?'}) (Confidence: 1.00, Lift: 17.67, Interest: 17.60)
Rule 4: If frozenset({'occupation_ ?', 'native-country_ United-States'}) then frozenset({'workclass_ ?'}) (Confidence: 1.00, Lift: 17.66, Interest: 17.59)
Rule 5: If frozenset({'workclass_ ?'}) then frozenset({'occupation_ ?', 'native-country_ United-States'}) (Confidence: 0.90, Lift: 17.66, Interest: 15.96)
Rule 6: If frozenset({'occupation_ ?'}) then frozenset({'workclass_ ?', 'native-country_ United-States'}) (Confidence: 0.90, Lift: 17.67, Interest: 15.90)
Rule 7: If frozenset({'workclass_ Private', 'relationship_ Own-child', 'sex_ Male', 'n

/usr/local/lib/python3.10/dist-packages/mlxtend/frequent_patterns/fpcommon.py:110: DeprecationWarning: DataFrames with non-bool types result in worse computationalperformance and their support might be discontinued in the future.Please use a DataFrame with bool type
  warnings.warn(


Rules for min_support=0.10 and min_interest=0.50:
Rule 1: If frozenset({'relationship_ Husband', 'race_ White', 'education_ HS-grad'}) then frozenset({'marital-status_ Married-civ-spouse', 'sex_ Male', 'native-country_ United-States'}) (Confidence: 0.95, Lift: 2.59, Interest: 2.46)
Rule 2: If frozenset({'sex_ Male', 'marital-status_ Married-civ-spouse', 'race_ White', 'education_ HS-grad'}) then frozenset({'relationship_ Husband', 'native-country_ United-States'}) (Confidence: 0.94, Lift: 2.59, Interest: 2.44)
Rule 3: If frozenset({'relationship_ Husband'}) then frozenset({'marital-status_ Married-civ-spouse', 'sex_ Male'}) (Confidence: 1.00, Lift: 2.44, Interest: 2.44)
Rule 4: If frozenset({'workclass_ Private', 'relationship_ Husband'}) then frozenset({'marital-status_ Married-civ-spouse', 'sex_ Male'}) (Confidence: 1.00, Lift: 2.44, Interest: 2.44)
Rule 5: If frozenset({'relationship_ Husband', 'race_ White'}) then frozenset({'marital-status_ Married-civ-spouse', 'sex_ Male'}) (Conf

/usr/local/lib/python3.10/dist-packages/mlxtend/frequent_patterns/fpcommon.py:110: DeprecationWarning: DataFrames with non-bool types result in worse computationalperformance and their support might be discontinued in the future.Please use a DataFrame with bool type
  warnings.warn(


In [ ]:
# c) Formulate questions and select attributes for association rule mining
# For example, let's say we want to find out rules related to high-income individuals who work in managerial positions and have a higher education level.

selected_attributes = ['education', 'occupation', 'income']

# Select relevant columns from the DataFrame
selected_df = df[selected_attributes]

# Binarize the selected DataFrame
selected_encoded_data = encoder.fit_transform(selected_df).toarray()
selected_encoded_df = pd.DataFrame(selected_encoded_data, columns=encoder.get_feature_names_out(selected_attributes))

# Calculate frequent itemsets and association rules
frequent_itemsets_selected = apriori(selected_encoded_df, min_support=0.1, use_colnames=True)
rules_selected = association_rules(frequent_itemsets_selected, metric="confidence", min_threshold=0.7)

# Display the top 10 rules
print("Top 10 rules based on selected attributes:")
for i, (_, row) in enumerate(rules_selected.nlargest(10, "confidence").iterrows(), start=1):
    antecedent = row['antecedents']
    consequent = row['consequents']
    confidence = row['confidence']
    lift = row['lift']
    print(f"Rule {i}: If {antecedent} then {consequent} (Confidence: {confidence:.2f}, Lift: {lift:.2f})")

Top 10 rules based on selected attributes:
Rule 1: If frozenset({'occupation_ Adm-clerical'}) then frozenset({'income_ <=50K'}) (Confidence: 0.87, Lift: 1.14)
Rule 2: If frozenset({'education_ HS-grad'}) then frozenset({'income_ <=50K'}) (Confidence: 0.84, Lift: 1.11)
Rule 3: If frozenset({'education_ Some-college'}) then frozenset({'income_ <=50K'}) (Confidence: 0.81, Lift: 1.07)


/usr/local/lib/python3.10/dist-packages/ipykernel/ipkernel.py:283: DeprecationWarning: `should_run_async` will not call `transform_cell` automatically in the future. Please pass the result to `transformed_cell` argument and any exception that happen during thetransform in `preprocessing_exc_tuple` in IPython 7.17 and above.
  and should_run_async(code)
/usr/local/lib/python3.10/dist-packages/mlxtend/frequent_patterns/fpcommon.py:110: DeprecationWarning: DataFrames with non-bool types result in worse computationalperformance and their support might be discontinued in the future.Please use a DataFrame with bool type
  warnings.warn(



The top 3 rules based on selected attributes are:


1.   If a person works in an administrative or clerical occupation (Adm-clerical), then they are likely to have an income of less than or equal to $50K (income_ <=50K) with a confidence of 0.87 and lift of 1.14.

2.   If a person's highest education level is high school graduation (HS-grad), then they are likely to have an income of less than or equal to $50K (income_ <=50K) with a confidence of 0.84 and lift of 1.11.

3.   If a person's highest education level is some college (Some-college), then they are likely to have an income of less than or equal to $50K (income_ <=50K) with a confidence of 0.81 and lift of 1.07.

These rules provide insights into the relationship between occupation and education level with income status. They suggest that individuals in administrative or clerical positions or with lower education levels (such as high school graduates or those with some college education) are more likely to have incomes below $50K. This insight could be useful for various purposes, such as targeted marketing strategies, understanding socioeconomic demographics, or designing educational and career development programs.

In [ ]:
# d) Try evaluation measures other than confidence and lift
# For example, we can try using leverage as an evaluation measure

rules_leverage = association_rules(frequent_itemsets_selected, metric="leverage", min_threshold=0.0)

# Display the top 10 rules based on leverage
print("Top 10 rules based on leverage:")
for i, (_, row) in enumerate(rules_leverage.nlargest(10, "leverage").iterrows(), start=1):
    antecedent = row['antecedents']
    consequent = row['consequents']
    confidence = row['confidence']
    leverage = row['leverage']
    print(f"Rule {i}: If {antecedent} then {consequent} (Confidence: {confidence:.2f}, Leverage: {leverage:.2f})")

Top 10 rules based on leverage:
Rule 1: If frozenset({'income_ <=50K'}) then frozenset({'education_ HS-grad'}) (Confidence: 0.36, Leverage: 0.03)
Rule 2: If frozenset({'education_ HS-grad'}) then frozenset({'income_ <=50K'}) (Confidence: 0.84, Leverage: 0.03)
Rule 3: If frozenset({'occupation_ Adm-clerical'}) then frozenset({'income_ <=50K'}) (Confidence: 0.87, Leverage: 0.01)
Rule 4: If frozenset({'income_ <=50K'}) then frozenset({'occupation_ Adm-clerical'}) (Confidence: 0.13, Leverage: 0.01)
Rule 5: If frozenset({'education_ Some-college'}) then frozenset({'income_ <=50K'}) (Confidence: 0.81, Leverage: 0.01)
Rule 6: If frozenset({'income_ <=50K'}) then frozenset({'education_ Some-college'}) (Confidence: 0.24, Leverage: 0.01)


/usr/local/lib/python3.10/dist-packages/ipykernel/ipkernel.py:283: DeprecationWarning: `should_run_async` will not call `transform_cell` automatically in the future. Please pass the result to `transformed_cell` argument and any exception that happen during thetransform in `preprocessing_exc_tuple` in IPython 7.17 and above.
  and should_run_async(code)



The top 6 rules based on leverage are:

1.   If a person has an income of less than or equal to $50K, then they are likely to have a high school graduation education level (HS-grad) with a confidence of 0.36 and leverage of 0.03.

2.   If a person's highest education level is high school graduation (HS-grad), then they are likely to have an income of less than or equal to $50K with a confidence of 0.84 and leverage of 0.03.

3.   If a person works in an administrative or clerical occupation (Adm-clerical), then they are likely to have an income of less than or equal to $50K with a confidence of 0.87 and leverage of 0.01.

4.   If a person has an income of less than or equal to $50K, then they are likely to work in an administrative or clerical occupation (Adm-clerical) with a confidence of 0.13 and leverage of 0.01.

5.   If a person's highest education level is some college (Some-college), then they are likely to have an income of less than or equal to $50K with a confidence of 0.81 and leverage of 0.01.

6.   If a person has an income of less than or equal to $50K, then they are likely to have some college education with a confidence of 0.24 and leverage of 0.01.

These rules provide insights into the relationship between income, education level, and occupation, specifically focusing on individuals with incomes of less than or equal to $50K.

Rule 1 and Rule 2 indicate a strong association between having an income of <=$50K and having a high school graduation education level (HS-grad). This insight could be valuable for understanding the demographics of individuals with lower incomes and potentially tailoring educational or economic policies to support this group.

Rule 3 and Rule 4 highlight the association between working in administrative or clerical occupations and having incomes <=$50K. This information could be useful for workforce planning, career counseling, or targeted job training programs.